src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.hadoop.fs.Path;
+import org.apache.parquet.avro.AvroParquetReader;
+import org.apache.parquet.avro.AvroParquetWriter;
+import org.apache.parquet.hadoop.ParquetFileWriter;
+import org.apache.parquet.hadoop.ParquetReader;
+import org.apache.parquet.hadoop.ParquetWriter;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+import org.apache.parquet.hadoop.util.HadoopInputFile;
+import org.apache.parquet.io.InputFile;
+import org.apache.parquet.io.OutputFile;
+
+public class ParquetReaderWriterWithAvro {
+
+  private ParquetReaderWriterWithAvro() {}
+
+  public static List<GenericData.Record> readFromParquet(InputFile inputFile) throws IOException {
+    List<GenericData.Record> records = new ArrayList<>();
+    try (ParquetReader<GenericData.Record> reader =
+        AvroParquetReader.<GenericData.Record>builder(inputFile).build()) {
+      GenericData.Record record;
+
+      while ((record = reader.read()) != null) {
+        records.add(record);
+      }
+      return records;
+    }
+  }
+
+  public static List<GenericData.Record> readFromParquet(Path filePath) throws IOException {
+    var config = new Configuration();
+    config.set("parquet.avro.readInt96AsFixed", "true");
+    InputFile inputFile = HadoopInputFile.fromPath(filePath, config);
+    return readFromParquet(inputFile);
+  }
+
+  public static void writeToParquet(
+      List<GenericData.Record> recordsToWrite,
+      List<String> columnNames,
+      Map<String, TableDataType> columnDataTypeMap,
+      OutputFile fileToWrite,
+      Schema schema,
+      Configuration config)
+      throws IOException {
+    try (ParquetWriter<GenericData.Record> writer =
+        AvroParquetWriter.<GenericData.Record>builder(fileToWrite)
+            .withSchema(schema)
+            .withConf(config)
+            .withCompressionCodec(CompressionCodecName.SNAPPY)
+            .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
+            .build()) {
+
+      for (GenericData.Record record : recordsToWrite) {
+        var newRecord = new GenericData.Record(schema);
+        for (var column : columnNames) {
+          switch (columnDataTypeMap.get(column)) {
+            case DATETIME, TIMESTAMP:
+              // Convert from fixed length binary to a long representing microseconds since epoch
+              GenericData.Fixed dtFixed = (GenericData.Fixed) record.get(column);
+              if (dtFixed == null) {
+                newRecord.put(column, null);
+              } else {
+                var bytes = dtFixed.bytes();
+                ByteBuffer bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
+                long timeOfDayNanos = bb.getLong();
+                var julianDay = bb.getInt();
+                // Given timeOfDayNanos and julianDay, convert to microseconds since epoch
+                Long microSeconds =
+                    (long) (julianDay - 2440588) * 86400 * 1000000 + timeOfDayNanos / 1000;
```

Comment:
You'll want to include a link to the source of this magic formula. Not sure where you got it from, I found it here: https://github.com/xhochy/parquet-format/blob/cb4727767823ae201fd567f67825cc22834c20e9/LogicalTypes.md#int96-timestamps-also-called-impala_timestamp

Reply:
Another reference: https://github.com/dremio/dremio-oss/blob/master/sabot/kernel/src/main/java/com/dremio/exec/store/parquet/ParquetReaderUtility.java#L70

Reply:
I've included the link in the code comments. Thanks for tracking down that first reference!